File: ThreatModelDfdService/Services/Impl/DfdElementService.cs
using ThreatModelDfdService.Data.Converter.Impl;
using ThreatModelDfdService.Data.DTO;
using ThreatModelDfdService.Model.Context;
using ThreatModelDfdService.Model.Entity;
using ThreatModelDfdService.Model.Enums;
using ThreatModelDfdService.Repositories;

namespace ThreatModelDfdService.Services;

public class DfdElementService(
    IRepository<DfdElement> dfdElementRepository,
    DfdElementMapper dfdElementMapper,
    MSSQLContext _context
)
{
    public async Task CreateOrUpdateAsync(long dfdId, UpsertDfdElementDTO dto)
    {
        if (dto.Id != null && dto.Id > 0)
        {
            DfdElement dbElement = GetById(dto.Id.Value);
            dbElement.Name = dto.Name;
            dbElement.XValue = dto.XValue;
            dbElement.YValue = dto.YValue;
            dbElement.Width = dto.Width;
            dbElement.Height = dto.Height;

            await dfdElementRepository.UpdateAsync(dbElement);
        }
        else
        {
            await CreateNewElementAsync(dfdId, dto);
        }
    }

    private async Task CreateNewElementAsync(long dfdId, UpsertDfdElementDTO dto)
    {
        DfdElement newEntity = dto.Type switch
        {
            DfdElementType.Process => new Process(),
            DfdElementType.Actor => new Actor(),
            DfdElementType.DataStore => new DataStore(),
            _ => throw new ArgumentException(
                "Dfd element type is not valid. | Name: " + dto.Name + " Type: " + dto.Type)
        };

        newEntity.DfdId = dfdId;
        newEntity.Name = dto.Name;
        newEntity.XValue = dto.XValue;
        newEntity.YValue = dto.YValue;
        newEntity.Width = dto.Width;
        newEntity.Height = dto.Height;
        newEntity.Type = dto.Type;
        newEntity.UuidIdentifier = dto.Uuid;

        await dfdElementRepository.CreateAsync(newEntity);
    }

    public List<DfdElementResponseDTO> GetDfdElementsByDfdId(long dfdId)
    {
        List<DfdElement> dfdElements = _context.DfdElements.Where(e => e.DfdId == dfdId).ToList();
        return dfdElementMapper.ParseList(dfdElements);
    }

    public DfdElement GetById(long id)
    {
        DfdElement dbElement = dfdElementRepository.FindById(id);
        ValidateDfdElementExists(dbElement);
        return dbElement;
    }

    public DfdElement GetByUUID(string uuid)
    {
        DfdElement dbElement = _context.DfdElements.FirstOrDefault(e => e.UuidIdentifier == uuid);
        ValidateDfdElementExists(dbElement);
        return dbElement;
    }

    public void DeleteDfdElement(long id)
    {
        DfdElement dbElement = GetById(id);
        dfdElementRepository.Delete(dbElement);
    }

    public void ValidateDfdElementExists(DfdElement dfdElement)
    {
        if(dfdElement == null) throw new KeyNotFoundException("Dfd element does not exist.");
    }
}

File: dfd-service/src/main/java/com/tcc/dfd_service/service/DfdElementService.java
package com.tcc.dfd_service.service;

import com.tcc.dfd_service.dto.DfdElementResponseDTO;
import com.tcc.dfd_service.dto.UpsertDfdElementDTO;
import com.tcc.dfd_service.entity.*;
import com.tcc.dfd_service.entity.Process;
import com.tcc.dfd_service.mapper.DfdElementMapper;
import com.tcc.dfd_service.repository.DfdElementRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class DfdElementService {

    private final DfdElementRepository dfdElementRepository;
    private final DfdElementMapper dfdElementMapper;

    @Transactional
    public void createOrUpdate(Long dfdId, UpsertDfdElementDTO dto) {

        if (dto.id() != null && dto.id() > 0) {

            DfdElement dbElement =
                    getById(dto.id());

            dbElement.setName(dto.name());
            dbElement.setXValue(dto.xValue());
            dbElement.setYValue(dto.yValue());
            dbElement.setWidth(dto.width());
            dbElement.setHeight(dto.height());

        } else {
            createNewElement(dfdId, dto);
        }
    }

    private void createNewElement(Long dfdId, UpsertDfdElementDTO dto) {
        DfdElement newEntity;

        switch (dto.type()) {
            case PROCESS -> newEntity = new Process();
            case ACTOR -> newEntity = new Actor();
            case DATA_STORE -> newEntity = new DataStore();
            default -> throw new IllegalArgumentException(
                    "Invalid DFD element type. Name: " + dto.name() + " Type: " + dto.type());
        }

        newEntity.setDfdId(dfdId);
        newEntity.setName(dto.name());
        newEntity.setXValue(dto.xValue());
        newEntity.setYValue(dto.yValue());
        newEntity.setWidth(dto.width());
        newEntity.setHeight(dto.height());
        newEntity.setType(dto.type());
        newEntity.setUuidIdentifier(dto.uuid());

        dfdElementRepository.save(newEntity);
    }

    public List<DfdElementResponseDTO> getDfdElementsByDfdId(Long dfdId) {
        List<DfdElement> elements = dfdElementRepository.findByDfdId(dfdId);
        return dfdElementMapper.parseList(elements);
    }

    public DfdElement getById(Long id) {
        return dfdElementRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("DFD element does not exist."));
    }

    public DfdElement getByUUID(String uuid) {
        return dfdElementRepository.findByUuidIdentifier(uuid)
                .orElseThrow(() -> new IllegalArgumentException("DFD element does not exist."));
    }

    @Transactional
    public void deleteDfdElement(Long id) {
        DfdElement dbElement = getById(id);
        dfdElementRepository.delete(dbElement);
    }
}